explorer/ast/value.h

```diff
@@ -95,6 +96,38 @@ class Value {
   const Kind kind_;
 };
 
+// Contains the result of the evaluation of an expression, including a value,
+// the original expression category, and an optional address if available.
+class ExpressionResult {
+ public:
+  ExpressionResult(Nonnull<const Value*> v, std::optional<Address> address,
+                   ExpressionCategory cat)
+      : value_(v), address_(std::move(address)), expr_cat_(cat) {}
+
+  static auto Value(Nonnull<const Carbon::Value*> v) -> ExpressionResult {
+    return ExpressionResult(v, std::nullopt, ExpressionCategory::Value);
+  }
+  static auto Reference(Nonnull<const Carbon::Value*> v, Address address)
+      -> ExpressionResult {
+    return ExpressionResult(v, std::move(address),
+                            ExpressionCategory::Reference);
+  }
+  static auto Initializing(Nonnull<const Carbon::Value*> v, Address address)
+      -> ExpressionResult {
+    return ExpressionResult(v, std::move(address),
+                            ExpressionCategory::Initializing);
+  }
```

Comment:
```suggestion
  static auto Value(Nonnull<const Carbon::Value*> v) -> ExpressionResult {
    return ExpressionResult(v, std::nullopt, ExpressionCategory::Value);
  }
  static auto Reference(Nonnull<const Carbon::Value*> v, Address address)
      -> ExpressionResult {
    return ExpressionResult(v, std::move(address),
                            ExpressionCategory::Reference);
  }
  static auto Initializing(Nonnull<const Carbon::Value*> v, Address address)
      -> ExpressionResult {
    return ExpressionResult(v, std::move(address),
                            ExpressionCategory::Initializing);
  }

  ExpressionResult(Nonnull<const Value*> v, std::optional<Address> address,
                   ExpressionCategory cat)
      : value_(v), address_(std::move(address)), expr_cat_(cat) {}
```

Per style, factory functions come before constructors (4 and 5 in [Declaration Order](https://google.github.io/styleguide/cppguide.html#Declaration_Order))

Reply:
Done